include/tarxx.h

```diff
@@ -292,18 +361,93 @@ namespace tarxx {
+            }
+
+            lz4_ctx_ = std::make_unique<lz4_ctx>();
+            const auto outbuf_size = lz4_check_error(LZ4F_compressBound, 16 * 1024, &lz4_prefs_);
+
+            lz4_out_buf_.reserve(outbuf_size);
+            const auto headerSize = lz4_check_error(LZ4F_compressBegin, lz4_ctx_->get(), lz4_out_buf_.data(), lz4_out_buf_.capacity(), &lz4_prefs_);
+            lz4_out_buf_pos_ += headerSize;
+            write_lz4_data();
+        }
+
+        void write_lz4_data()
+        {
+            unsigned long long offset = 0;
+            switch (mode_) {
+                // guarding that input is not stream is done in add_file_streaming
+                // using a file as input is fine, as we have all data necessary
+                // for compression w/o seeking
+                case output_mode::stream_output:
+                    while (lz4_out_buf_pos_ > 0) {
+                        block_t block {};
+                        const auto copy_size = std::min(lz4_out_buf_pos_, block.size());
+                        std::copy_n(lz4_out_buf_.begin() + offset, copy_size, block.data());
+                        callback_(block, copy_size);
+                        lz4_out_buf_pos_ -= copy_size;
+                        offset += copy_size;
+                    }
+
+                    break;
+                case output_mode::file_output:
+                    file_.write(lz4_out_buf_.data(), lz4_out_buf_pos_);
+                    lz4_out_buf_pos_ = 0;
+                    break;
+            }
+        }
+
         enum class output_mode : unsigned {
             file_output,
             stream_output
         };
 
+        class lz4_ctx {
+        public:
+            lz4_ctx()
+            {
+                LZ4F_createCompressionContext(&ctx_, LZ4F_VERSION);
+            }
+            ~lz4_ctx()
+            {
+                free(ctx_);
+            }
+
+            LZ4F_cctx* get()
+            {
+                return ctx_;
+            }
+
+        private:
+            LZ4F_compressionContext_t ctx_ = nullptr;
+        };
+
         tar_type type_;
         output_mode mode_;
+        compression_mode compression_;
         std::fstream file_;
         callback_t callback_;
         long stream_file_header_pos_;
         block_t stream_block_;
         size_t stream_block_used_;
+
+        std::unique_ptr<lz4_ctx> lz4_ctx_;
+        std::vector<char> lz4_out_buf_;
+        size_t lz4_out_buf_pos_ = 0;
+
+        static inline constexpr LZ4F_preferences_t lz4_prefs_ = {
+                {LZ4F_max256KB, LZ4F_blockIndependent, LZ4F_noContentChecksum,
+                 LZ4F_frame, 0 /* unknown content size */, 0 /* no dictID */,
+                 LZ4F_noBlockChecksum},
+                0,         /* compression level; 0 == default */
+                0,         /* autoflush */
+                0,         /* favor decompression speed */
+                {0, 0, 0}, /* reserved, must be set to 0 */
+        };
```

Comment:
please only compile in if lz4 support is activated (see comments in `CMakeLists`)